### ISSUE TITLE
Unable to program PIC12F683.

### ISSUE THREAD

**Issue by DavidGriffith** [DavidGriffith/minipro-import-test#267]
In GitLab by @virtadpt on Jun 19, 2022, 16:40

I'm trying to program some PIC12F683's with pre-compiled firmware ([https://projectmf.org/downloads/BBsoftware.zip](https://projectmf.org/downloads/BBsoftware.zip) - filename ADC_blue_box_16_sp.hex in the archive) and so far I haven't been able to get a working PIC when using a TL866CS programmer.  I'm seeing some unusual things that might be a bug in minipro, might be a problem in my programmer, or might be a combination of the two.

I've tried two different batches of PIC12F683's so far, one group sourced from Amazon and one group sourced from Mouser.  They react the same way.

(note: unless otherwise noted, the results are the same for chips in either batch.)

The version of Minipro I'm using:

```
{16:21:26 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --version
Supported programmers: TL866A/CS, TL866II+
Found TL866CS 03.2.86 (0x256)
minipro version 0.5     A free and open TL866XX programmer
Commit date:	2022-03-25 18:28:45 -0700
Git commit:	fefd16015aa6b6900d8550e59dafc66df96775b6
Git branch:	master
TL866A/CS:	14330 devices, 9 custom
TL866II+:	16442 devices, 10 custom
```

From first principles, minipro detects my programmer correctly:

```
{16:25:01 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --presence_check
tl866a: TL866CS
```

Minipro detects my PICs correctly:

```
{16:25:26 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --search 12F683
Found TL866CS 03.2.86 (0x256)
PIC12F683
PIC12F683@DFN8
PIC12F683@SOIC8
```

For some reason, pin testing doesn't seem to work:

```
{16:25:31 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device PIC12F683 --pin_check
Found TL866CS 03.2.86 (0x256)
Pin test is not supported.
```

I can, however, query the chips' specifics reliably:

```
{16:27:13 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --get_info PIC12F683
Found TL866CS 03.2.86 (0x256)
Name: PIC12F683
Memory: 2048 Words + 256 Bytes
Package: DIP8
ICSP: ICP002.JPG
Protocol: 0x63
Read buffer size: 128 Bytes
Write buffer size: 32 Bytes
```

As will become relevant in a few paragraphs, `--get_info` printing the chip ID would be helpful here.

If I try to erase it I get an invalid chip ID:

```
{16:27:18 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device PIC12F683 --erase
Found TL866CS 03.2.86 (0x256)
Invalid Chip ID: expected 0x0023, got 0x0520 (unknown)
(use '-y' to continue anyway at your own risk)

{16:30:37 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device PIC12F683 --erase -y
Found TL866CS 03.2.86 (0x256)
WARNING: Chip ID mismatch: expected 0x0023, got 0x0029 (unknown)
Erasing... 0.03Sec OK
```

Minipro gets two different chip IDs depending on whether or not I use `-y` to force an erasure.  The erasure does seem to take, though:

```
{16:33:44 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device pic12f683 --read - -y | xxd -b -c 1 | less
00000000: 00000000  .
00000001: 00000000  .
00000002: 00000000  .
...
00000ffd: 00000000  .
00000ffe: 10100100  .
00000fff: 00110100  4
```

Here's what happens when I try to flash the .hex image:

```
{16:34:16 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device pic12f683 --write ADC_blue_box_16_sp.hex 
Found TL866CS 03.2.86 (0x256)
Invalid Chip ID: expected 0x0023, got 0x0520 (unknown)
(use '-y' to continue anyway at your own risk)

{16:35:02 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device pic12f683 --write ADC_blue_box_16_sp.hex -y
Found TL866CS 03.2.86 (0x256)
WARNING: Chip ID mismatch: expected 0x0023, got 0x0029 (unknown)
Found Intel hex file.
Erasing... 0.03Sec OK
Writing Code...  1.79Sec  OK
Reading Code...  0.18Sec  OK
Verification failed at address 0x0000: File=0x2946, Device=0x0000
```

Again, two different chip IDs.  If I dump the contents of the PIC again, there's no change:

```
{16:36:30 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device pic12f683 --read - -y | xxd -b -c 1 | less
00000000: 00000000  .
00000001: 00000000  .
00000002: 00000000  .
...
00000ffd: 00000000  .
00000ffe: 10100100  .
00000fff: 00110100  4
```

I have similar problems with a batch of PIC16F628A's for a different project, but that should go in a separate bug ticket to avoid muddying the waters needlessly.

In a couple of other bug reports here, I've read that it could be a bad electrical connection.  I'm waiting on a batch of ZIF sockets to arrive so I can swap it out but as of right now I've tried various methods of cleaning the contacts on both the chips and existing ZIF socket, and none of them have helped.  I've also tried jiggling and scraping the chip sideways in the existing ZIF, to no avail.

I'm a little out of my depth here because I'm trying to learn how to do microcontroller programming in general (and PIC programming in particular), which probably doesn't help the situation.  Am I doing something dumb because I don't fully understand how PICs and chip programmers work, or is there something more going on?

**Comment by DavidGriffith** [comment]
In GitLab by @virtadpt on Jun 19, 2022, 16:49

Going line by line through the online help for minipro, here's what comes back for `--read_id`:

```
{16:44:37 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device PIC12F683 --read_id
Found TL866CS 03.2.86 (0x256)
Chip ID mismatch: expected 0x0023, got 0x0029 (unknown)
```

`--logic_test` doesn't seem to work, either:

```
{16:48:32 @ Sun Jun 19}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device PIC12F683 --logic_test
Found TL866CS 03.2.86 (0x256)
TL866CS: logic IC test not implemented
```

It's possible that there are programming options (for `--pulse` and `--vpp`) but I don't know of any, and there don't seem to be any documented, so I haven't messed with those CLI flags.

Using `--hardware_check` says that all of the VPP, VCC, and GND driver pins are okay.  Same with overcurrent protection.  I can post the output but it's probably not helpful.

**Comment by DavidGriffith** [comment]
In GitLab by @virtadpt on Jun 27, 2022, 13:05

Is this not actually a bug?  PEBKAC?

**Comment by DavidGriffith** [comment]
Sorry, I've been out of things for a while.  I haven't had a chance to take a good look at this or PICs in general to decide if this is a PEBKAC or not.

If you're really keen to fiddle with a bluebox implemented with a microcontroller, look at https://gitlab.com/DavidGriffith/bluebox-esquire.  I created that and the firmware within as a sort of reimplementation of Don Froula's work, which is what you're playing with.  Rather than a PIC, my box uses an AVR.  I can confirm that my TL866CS and the minipro program will program the ATtiny85 for this without trouble.  Until I can get that project updated, here's a script I used to program the AVR:

````
minipro -p attiny85 -w bluebox.hex -c code
minipro -e -p attiny85 -w fuses.txt -c config
minipro -e -p attiny85 -w bluebox.eep.hex -c data
````

The file `fuses.txt`:
````
fuses_lo = 0xff
fuses_hi = 0xdb
fuses_ext = 0xff
lock_byte = 0xff
````

Minipro really should support specifying fuses at the command line...  Something for me to add soon, I suppose.

**Comment by DavidGriffith** [comment]
In GitLab by @radiomanV on Jun 27, 2022, 23:53

Sorry for the delay, i'm quite busy these days.  
 
Several things need to be clarified here.   
Because i'm one of the guys who implemented some of the involved code i'll try to answer to your questions.   

First the `--pin_check` is not implemented in TL866A/CS programmers because those programmers hardware doesn't have the required hardware support (pull-up/pull-down resistor for each i/o line). So this is only implemented in TL866II+ and higher models (T48/T56) so the `Pin test is not supported.` message is correct. Yes in your case you don't know if the chip make a good socket contact or not.   

The `--logic_test` is not implemented because again some hardware features are missing in TL866A/CS models, but the logic test is for testing logic chips like TTL 7400 or CMOS 4000 series not for what you probably think.   

The `--hardware_check` as you probably guess is for checking the pin drivers (a bunch of npn/pnp transistors used to power the chip with GND/VCC/VPP) so this is a good indicator that the chip can be powered correctly.   

Now to the issue. The PIC controllers can be identified by a word called "device ID" and depending on the chip this word can have a different length. For 12F6XX/16F6XX (and others) this word is 14 bit length. The most 9 MSB bits (bit 13 to 5) is the actual chip ID and the last LSB 5 bits (bit 4 to 0) are the so called "revision bits".   
Look at this table:   
```
Device      Dev ID    Revision
PIC12F635 00 1111 101 x xxxx
PIC12F683 00 0100 011 x xxxx (0x460 or 0x23 + 00000)
PIC16F631 01 0100 001 x xxxx
PIC16F636 01 0000 101 x xxxx
PIC16F639 01 0000 101 x xxxx
PIC16F677 01 0100 010 x xxxx
PIC16F684 01 0000 100 x xxxx
PIC16F685 00 0100 101 x xxxx
PIC16F687 01 0011 001 x xxxx
PIC16F688 01 0001 100 x xxxx
PIC16F689 01 0011 010 x xxxx
PIC16F690 01 0100 000 x xxxx
```
When comparing the chip ID we shift right by 5 bits the whole ID word thus throwing the revision number away and comparing the read ID with the database ID (in infoic.xml). So we only compare the ID bits (first 9 MSB bits) because we can have different chip revisions for the same chip. The revision bits (bit 4 to 0) are useful only when printing the chip ID and is not involved in any comparison.   
The `Invalid Chip ID: expected 0x0023, got 0x0520 (unknown)` is a bug in error printing routine (forgot a shift) but this is only a cosmetic thing easily fixable in `main.c` file line 2551:
```c
// Instead of this
 fprintf(
              stderr,
              "Invalid Chip ID: expected 0x%04X, got 0x%04X (%s)\n(use '-y' "
              "to continue anyway at your own risk)\n",
              handle->device->chip_id >> shift, chip_id_temp,
              name ? name : "unknown");
// Should be this
 fprintf(
              stderr,
              "Invalid Chip ID: expected 0x%04X, got 0x%04X (%s)\n(use '-y' "
              "to continue anyway at your own risk)\n",
              handle->device->chip_id >> shift, chip_id_temp >> shift,
              name ? name : "unknown");
```
So this shouldn't be a problem but because `0x29` (or 0x520 unshifted) chip ID is not found in the database we mark this as `unknown`. Can be a firmware bug? possible, or a bad pin contact? well, judging that the chip cannot be programmed i think yes, it can be but i'm not sure.    
Searching some personal databases i found that the 0x520/0x29 ID belongs to the PIC18F4320 (which is not supported) so clearly something is fishy here.   
> I've tried two different batches of PIC12F683's so far, one group sourced from Amazon and one group sourced from Mouser. They react the same way.   

Well this is weird. Can you test the Windows software?(it can be executed under Linux via wine with a special driver). The culprit must be identified (chip/software/firmware/programmer) in order to search for a fix.

**Comment by DavidGriffith** [comment]
In GitLab by @virtadpt on Jun 29, 2022, 13:31

`--pin_check` not implemented... okay, that rules out a couple of corner cases, plus it's good to know about this particular programmer.  Same with `--logic_check`, though now that I know what it's for I can play around with that feature using some other stuff in my parts bin.

I can give the Windows software a shot this weekend.  Which driver would be required for Wine?  I was going to build a Windows VM and have an agent track down the Windows software for me because it didn't come with anything (serves me right for buying something off of Amazon, I guess).

**Comment by DavidGriffith** [comment]
In GitLab by @virtadpt on Jun 29, 2022, 13:32

I'll give it a try, thank you.

**Comment by DavidGriffith** [comment]
In GitLab by @radiomanV on Jun 29, 2022, 15:09

Here https://github.com/radiomanV/TL866/tree/master/wine is a special usb wrapper/driver for wine. Just install wine, install the windows software and copy the provided `setupapi.dll` file in the xgpro folder. The udev rule is already installed so you can start the windows software. If you have any difficulties please open a new issue there.

**Comment by DavidGriffith** [comment]
In GitLab by @virtadpt on Jul 10, 2022, 16:23

I just tried your suggestion and every function I've tried in the software (blank check, read IC, erase, blank check, read ID, and so forth) keep returning "Programmer Device is Unavailable or Error."

On startup, I'm selecting TL866-II Plus as the programmer.  Same thing when I select TL866-3G as the programmer.

This could be PEBKAC or it could be a that I have a crap programmer.

**Comment by DavidGriffith** [comment]
In GitLab by @virtadpt on Jul 10, 2022, 22:08

I applied your patch to `main.c` (minipro commit cf0fcb0575096a2b40e0fb33707cf4da91a72eee), recompiled, and reinstalled.

If I erase the chip it seems to work:

```
{21:59:22 @ Sun Jul 10}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device PIC12F683 --erase
Found TL866CS 03.2.86 (0x256)
Chip ID OK: 0x0023 Rev.0x03
Erasing... 0.03Sec OK
```

Dumping the contents of the chip seems to show that it's empty:

```
{22:01:04 @ Sun Jul 10}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device PIC12F683 --read - -y | xxd -b -c 1 | less
Found TL866CS 03.2.86 (0x256)
Chip ID OK: 0x0023 Rev.0x03
Reading Code...  0.19Sec  OK
00000000: 11111111  .
00000001: 00111111  ?
00000002: 11111111  .
00000003: 00111111  ?
...
```
...and doing a blank check does not return an error:

```
{22:02:04 @ Sun Jul 10}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device PIC12F683 --blank_check
Found TL866CS 03.2.86 (0x256)
Chip ID OK: 0x0023 Rev.0x03
Reading Code...  0.18Sec  OK
Code memory section is blank.
Reading Data...  0.02Sec  OK
Data memory section is blank.
```

Writing the .hex image seems to work now.

```
{22:02:10 @ Sun Jul 10}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device PIC12F683 --write ADC_blue_box_16_sp.hex
Found TL866CS 03.2.86 (0x256)
Chip ID OK: 0x0023 Rev.0x03
Found Intel hex file.
Erasing... 0.03Sec OK
Writing Code...  1.80Sec  OK
Reading Code...  0.18Sec  OK
Verification OK
```

...and dumping its contents shows something different:

```
{22:03:22 @ Sun Jul 10}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device PIC12F683 --read - -y | xxd -b -c 1 | less
Found TL866CS 03.2.86 (0x256)
Chip ID OK: 0x0023 Rev.0x03
Reading Code...  0.19Sec  OK
00000000: 01000110  F
00000001: 00101001  )
00000002: 10100000  .
00000003: 00000000  .
00000004: 01111111  .
00000005: 00111001  9
00000006: 00100000   
00000007: 00011011  .
00000008: 10000000  .
00000009: 00111100  <
...
```

But when I replace the chip on the board and test it, it still doesn't work.

At this point, I'm not sure what's going on.  Could be my programmer, could be the firmware images themselves (I tried both .hex files).  I am, unfortunately, at the limit of my knowledge.

**Comment by DavidGriffith** [comment]
In GitLab by @radiomanV on Jul 11, 2022, 03:11

It looks like the minipro can write and read back the written data correctly. There's a compare mask which masks every word to 0x3fff. It looks ok for me. For example this:
```
00000000: 01000110  F // 0x46
00000001: 00101001  ) // 0x29
```
So the first word is 0x2946 and if we look at the hex file it's correct.  
:10000000**4629**A0007F39201B803C0920A01B003C12   

I think that you have to program your fusebits manually. Unfortunately minipro is yet unable to extract those config/fusebits from the hex file. For now you must supply the fusebits manually as a file (@DavidGriffith is right, we should implement some command line arguments).  
Looking at the hex file the config word is this:
```
:02400E00E2309E // 0x30E2
```
So after the hex file is written ok, please read the config word to a file like this:
` ./minipro --device PIC12F683 --page config --read ./config`. Then open the config file and modify the `conf_word = ...` with `conf_word = 0x30E2`    
Then write the config word with `./minipro --device PIC12F683 --page config --write ./config`. The chip erase is skipped when fusebits are written so that's why we should write them in last step. Of course the `-e` can be passed to skip the erase and then we can write in any order we want.  
This is a bit odd but it works.

**Comment by DavidGriffith** [comment]
In GitLab by @radiomanV on Jul 11, 2022, 03:22

> "Programmer Device is Unavailable or Error."

Unfortunately your programmer is not detected. If everything is ok you should be able to use your TL866 under Linux. Pay attention because if you use a 64 bit distro then you should install the 32 bit libusb and libudev because the minipro/xgpro softwares are 32 bit applications!     

You can run minipro/xgpro from the command line with `wine ./Xgpro.exe` to see more debug messages. When loaded correctly my `setupapi.dll` wrapper will print some debug messages. If this library fails to load then you will be unable to use your programmer.     

Also please open a new issue here: https://github.com/radiomanV/TL866/issues We don't want to pollute this issue thread with off-topic things.

**Comment by DavidGriffith** [comment]
In GitLab by @virtadpt on Jul 11, 2022, 19:59

Okay, I've read that through a few times and that makes sense.  I was also able to dump the config page (thanks for that, by the way, I'd been dumping the entire chip and tinkering with the files separately).  Config page dump edited and re-flashed.  I also took a quick peek at the page from the chip, and it looks like the fuses were set correctly:

```
{19:50:41 @ Mon Jul 11}
[drwho @ windbringer projectMF_bluebox_firmware] () $ minipro --device pic12f683 --page config --read -
Found TL866CS 03.2.86 (0x256)
Chip ID OK: 0x0023 Rev.0x03
Reading fuses... 0.06Sec  OK
conf_word = 0x30e2
user_id0 = 0x3fff
user_id1 = 0x3fff
user_id2 = 0x3fff
user_id3 = 0x3fff
```

After reseating the chip in the socket and powering up, it's acting the same as before.  Unresponsive for a few seconds, then the same buzz over and over again.  And the battery lead just snapped at the PCB, so that's about all the tinkering I can do until I have time to turn my desk into a workbench again.  For what it's worth, I agree with David - implementing fusebit support would be essential.

**Comment by DavidGriffith** [comment]
In GitLab by @virtadpt on Jul 11, 2022, 20:01

I had a pack of the PCBs fabbed by PCBway and have one populated with the components I have on hand, but until the attiny85's come in (they're backordered - semiconductor shortage) I can't do much with them.

**Comment by DavidGriffith** [comment]
In GitLab by @virtadpt on Jul 11, 2022, 20:02

I do have the 32-bit libusb package installed.  It was in the documentation at the repo you pointed me to so I made sure to get that done at the same time that I installed Wine (I did a full system upgrade and then installed them, for whatever that's worth).  However, I agree and I'll take discussion of TL866/ over there.
